File: script/check-account-bulling/check-account-bulling.py
# コメント追加するで2222
# AWS Organizations内の全アカウントの請求情報を取得し、特定の基準を超えていたら通知する。
# 通知先はSNSトピックを指定する。
# 基準は毎月の予算の半分を超えた場合と予算を超えた場合に通知する。
# 予算はAWS Budgetsで設定する。
# 予算は各アカウント毎に毎月5万円とする。
import boto3
import os

def lambda_handler(event, context):
    # AWS Organizationsのクライアントを生成
    org = boto3.client('organizations')
    # AWS Organizationsの全アカウントを取得
    accounts = org.list_accounts()
    # AWS Budgetsのクライアントを生成
    budgets = boto3.client('budgets')
    # 予算の基準を取得
    # FIX: 予算の基準はbudgetから取得する
    budget_limit = 50000
    # 予算の基準を超えた場合に通知するSNSトピックを取得
    ses = boto3.client('ses')
    # 予算の基準を超えた場合に通知する
    # FIX: アカウントの使用量を取得する処理が必要。
    for account in accounts['Accounts']:
        # アカウントの予算を取得
        budget = budgets.describe_budget(
            AccountId=account['Id'],
            BudgetName='MonthlyBudget'
        )
        # 予算の基準を超えた場合は通知する（ここではSESを使う）

        if budget['Budget']['BudgetLimit'] > budget_limit:
            ses.send_email(
            Source='sugimori@sugichori.com',
            Destination={
                'ToAddresses': ['ryusugi777@gmail.com']
            },
            Message={
                'Subject': {
                'Data': 'Budget Exceeded'
                },
                'Body': {
                'Text': {
                    'Data': 'The budget of account ' + account['Id'] + ' has exceeded the limit'
                }
                }
            }
            )
        # 予算の半分を超えた場合は通知する（ここでもSESを使う）
        elif budget['Budget']['BudgetLimit'] / 2 > budget_limit:
            ses.send_email(
            Source='sugimori@sugichori.com',
            Destination={
                'ToAddresses': ['ryusugi777@gmail.com']
            },
            Message={
                'Subject': {
                'Data': 'Budget Exceeded'
                },
                'Body': {
                'Text': {
                    'Data': 'The budget of account ' + account['Id'] + ' has exceeded half of the limit'
                }
                }
            }
            )

File: script/check-account-bulling/check-account-bulling.go
package main

import (
	"context"
	"fmt"
	"log"
	"os"
	"strconv"

	"github.com/aws/aws-lambda-go/lambda"
	"github.com/aws/aws-sdk-go/aws"
	"github.com/aws/aws-sdk-go/aws/session"
	"github.com/aws/aws-sdk-go/service/budgets"
	"github.com/aws/aws-sdk-go/service/organizations"
	"github.com/aws/aws-sdk-go/service/sns"
)

func handler(ctx context.Context) {
	// AWSセッションを作成a
	sess := session.Must(session.NewSession())

	// AWS Organizationsのクライアントを生成
	org := organizations.New(sess)

	// AWS Organizationsの全アカウントを取得
	accounts, err := org.ListAccounts(&organizations.ListAccountsInput{})
	if err != nil {
		log.Fatalf("Failed to list accounts: %v", err)
	}

	// AWS Budgetsのクライアントを生成
	budgetsClient := budgets.New(sess)

	// 予算の基準を取得
	budgetLimit := 50000.0

	// SNSクライアントを生成
	snsClient := sns.New(sess)

	for _, account := range accounts.Accounts {
		// アカウントの予算を取得
		budget, err := budgetsClient.DescribeBudget(&budgets.DescribeBudgetInput{
			AccountId:  aws.String(*account.Id),
			BudgetName: aws.String("MonthlyBudget"),
		})
		if err != nil {
			log.Printf("Failed to describe budget for account %s: %v", *account.Id, err)
			continue
		}
		if budget.Budget.BudgetLimit != nil && budget.Budget.BudgetLimit.Amount != nil {
			amountStr := *budget.Budget.BudgetLimit.Amount
			amount, err := strconv.ParseFloat(amountStr, 64)
			if err != nil {
				log.Printf("Failed to parse budget amount for account %s: %v", *account.Id, err)
				continue
			}

			// 予算の基準を超えた場合は通知する
			if amount > budgetLimit {
				sendNotification(snsClient, *account.Id, "Budget Exceeded", "The budget of account "+*account.Id+" has exceeded the limit")
			} else if amount/2 > budgetLimit {
				sendNotification(snsClient, *account.Id, "Budget Exceeded", "The budget of account "+*account.Id+" has exceeded half of the limit")
			} else {
				log.Printf("No budget limit set for account %s", *account.Id)
			}

		}
	}
}

func sendNotification(snsClient *sns.SNS, accountId, subject, message string) {
	_, err := snsClient.Publish(&sns.PublishInput{
		TopicArn: aws.String(os.Getenv("SNS_TOPIC_ARN")),
		Message:  aws.String(message),
		Subject:  aws.String(subject),
	})
	if err != nil {
		log.Printf("Failed to send notification for account %s: %v", accountId, err)
	} else {
		fmt.Printf("Notification sent for account %s: %s\n", accountId, message)
	}
}

func main() {
	lambda.Start(handler)
}
